api/pkg/controller/addoninstaller/controller.go

```
@@ -0,0 +1,248 @@
+	if err == nil {
+		// Forget about the #AddRateLimited history of the key on every successful synchronization.
+		// This ensures that future processing of updates for this key is not delayed because of
+		// an outdated error history.
+		c.queue.Forget(key)
+		return
+	}
+
+	// This controller retries 5 times if something goes wrong. After that, it stops trying.
+	if c.queue.NumRequeues(key) < 5 {
+		glog.V(0).Infof("Error syncing %v: %v", key, err)
+
+		// Re-enqueue the key rate limited. Based on the rate limiter on the
+		// queue and the re-enqueue history, the key will be processed later again.
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	c.queue.Forget(key)
+	// Report to an external entity that, even after several retries, we could not successfully process this key
+	utilruntime.HandleError(err)
+	glog.V(0).Infof("Dropping %q out of the queue: %v", key, err)
+}
+
+// make sure that all default addons are installed on cluster creation
+func (c *Controller) sync(key string) error {
+	clusterFromCache, err := c.clusterLister.Get(key)
+	if err != nil {
+		return fmt.Errorf("failed to get cluster for the key: %s: %v", key, err)
+	}
+
+	cluster := clusterFromCache.DeepCopy()
+	if cluster.Labels[kubermaticv1.WorkerNameLabelKey] != c.workerName {
+		glog.V(8).Infof("skipping cluster %s due to different worker assigned to it", key)
+		return nil
+	}
+	// Reconciling
+	if !cluster.Status.Health.Apiserver {
+		glog.V(8).Infof("reconciling: Health.Apiserver of the cluster %s is false", key)
+		return nil
+	}
+
+	for _, defaultAddon := range c.defaultAddonList {
+		addonFromCache, err := c.addonLister.Addons(cluster.Status.NamespaceName).Get(defaultAddon)
+		if err != nil && kerrors.IsNotFound(err) {
+			c.createDefaultAddon(defaultAddon, cluster)
+		} else if err != nil {
+			utilruntime.HandleError(fmt.Errorf("failed to get addon %s for cluster %s: %v", defaultAddon, addonFromCache.Spec.Cluster.Name, err))
```

Return the error.
As with https://github.com/kubermatic/kubermatic/pull/1431/files#r201024582 we have a error handler in place.


But do not include the cluster name inside the error. I will get added from the error handler

---

we could return an error but what does it give us ? 
perhaps it is better to continue looping and try to install as much `addons` as we can ? Then at the end we could actually return an error indicating that the cluster should be added to the queue.

---

done